[WIP]feat: add git integration

alexandercbooth opened the pull request:
Hi All!

First -- sorry for the long delay 😞; I have missed you all. I recently discovered a new great git [project](https://isomorphic-git.github.io/) that I think could be a good fit for us, so I decided to take another spin at this for #2016. 

## In this initial _rough_ draft
- current branch is displayed in the status bar
![image](https://user-images.githubusercontent.com/15242567/40156690-0ce7fa50-5969-11e8-9051-129c511d93b9.png)
- some base funtionality added through a few menu items including: init, add, remove and commit
![image](https://user-images.githubusercontent.com/15242567/40156694-1ac66058-5969-11e8-8713-852136f25321.png)
![image](https://user-images.githubusercontent.com/15242567/40156699-1ffccb02-5969-11e8-8972-993de44de40f.png)

### Issues
- I have added a default git user for the commits as a place holder -- unless a local git config is created, the commits are not possible without a default user. We could create config on the fly, but this runs into an issue that crops up often: how to do a user prompt with electron (in my brief look around, I am not sure). 
- similarly for commit message-- at the moment, just defaults to "Notebook Updated" (I know, I know, _horrid_ commit practices with a message that dull 😅). We need a way for the user to enter a custom message. Same thing goes for branching (although I did not implement that yet for this reason).

My main question is what is the base functionality we want from this?

Looking forward to any/all feedback 😄

rgbkrk commented:
I love that you kept hamster.

rgbkrk commented:
I think it's totally fine to make the tradeoff of making a dummy commit message for now. We can always introduce a UI for filling it in later.

rgbkrk commented on applications/desktop/src/main/menu.js:
```diff
@@ -478,6 +478,32 @@ export function loadFullMenu(store: * = global.store) {
+        click: createSender("menu:git:remove")
+      }
+    ]
+  };
```
I can't tell if I like that there's a full Git Menu or if it should be grouped under another menu. I'll come back to my thinking on this.

rgbkrk commented on applications/desktop/src/notebook/menu.js:
```diff
@@ -368,6 +370,67 @@ export function dispatchRestartClearAll(
+      author: {
+        name: "Theo Decker",
+        email: "goldfinch@met.com"
+      },
```
Hmmm... Do we have to specify these or can this git binding autodetect the name / email from their gitconfig? If it's not defined are we able to help them set it up?

alexandercbooth replied:
Absolutely, it can autodetect from the local .git/config file currently (in the future, they plan to scan for a global config, but for now it's just the local). We can definitely help them set it up. I'd ideally like to have a config option under the menu, but with not being able to do prompts currently maybe it would have to be just an informational one that shows what's needed and where to put it. Or maybe it could put a template in the right place and direct them to add their name/email?

rgbkrk commented on applications/desktop/src/notebook/menu.js:
```diff
@@ -368,6 +370,67 @@ export function dispatchRestartClearAll(
+  const notebookdir = filepath.split("/");
+  const notebook = notebookdir[notebookdir.length - 1];
+  git.add({ ...repo, filepath: notebook });
+}
```
Each of these git operations return a promise. I'd love to get these into epics that we work with, both for sending off the initial `addToGitAction` as well as responses. Same applies to `init`, etc. as we also need to handle their failure cases.

alexandercbooth replied:
Perfect! I knew this was not the right way to do it 😅 , I will work on trying to change this tonight.

alexandercbooth commented:
Thanks @rgbkrk! Really excited to be able to work with you all again 😄.

alexandercbooth commented:
### Update
- removed the placeholder user
- git operations are now all epics
- added notifications
- added a menu option to copy a users global .gitconfig to the local repo
- moved the branch determination to a selector (not really sure where this should go exactly, but it didn't seem right in the status bar file 🤔)

rgbkrk commented:
Sometimes, sadly, you have to make sure you're updated with the remote repo and run `git clean -xfd && npm i` to make sure the dependencies all match up nicely. 😭

alexandercbooth commented:
Ah that did it, thanks!

rgbkrk commented:
Alex I'm going to close this since it's way out of date with the current repo. Feel free to re-open again.

It might be alright if you make some of the components and what-not separate PRs that we expose in the components docs. We can refine UI / UX without having to touch desktop bits for now.
